// @stackbit/cms-core/dist/utils/asset-sources-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAssetSourceBySourceName = exports.getImageFieldsFromSourceData = exports.transformAssetSourceDataForAssetSource = exports.getAssetSourcesForClient = void 0;
const BUILT_IN_ASSET_SOURCES = ['cloudinary', 'bynder', 'aprimo'];
function getAssetSourcesForClient(stackbitConfig) {
    var _a;
    if (!stackbitConfig) {
        return [];
    }
    const assetSources = (_a = stackbitConfig.assetSources) !== null && _a !== void 0 ? _a : [];
    return assetSources.reduce((accum, assetSource) => {
        var _a;
        if (assetSource.type === 'iframe') {
            if (!assetSource.name || !assetSource.url) {
                return accum;
            }
            const { transform, preview, ...rest } = assetSource;
            return accum.concat(rest);
        }
        else {
            const name = (_a = assetSource.name) !== null && _a !== void 0 ? _a : (BUILT_IN_ASSET_SOURCES.includes(assetSource.type) ? assetSource.type : null);
            if (!name) {
                return accum;
            }
            const { transform, preview, ...rest } = assetSource;
            return accum.concat({
                ...rest,
                name: name
            });
        }
        return accum;
    }, []);
}
exports.getAssetSourcesForClient = getAssetSourcesForClient;
function transformAssetSourceDataForAssetSource(sourceData, assetSource) {
    if (typeof assetSource.transform === 'function') {
        return assetSource.transform({ assetData: sourceData });
    }
    else {
        return sourceData;
    }
}
exports.transformAssetSourceDataForAssetSource = transformAssetSourceDataForAssetSource;
function getImageFieldsFromSourceData({ sourceData, imageModelField, assetSources }) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const imageAssetSource = getAssetSourceBySourceName(assetSources, imageModelField.source);
    if (imageAssetSource === null || imageAssetSource === void 0 ? void 0 : imageAssetSource.preview) {
        const preview = typeof imageAssetSource.preview === 'function' ? imageAssetSource.preview({ assetData: sourceData }) : imageAssetSource.preview;
        return {
            title: {
                type: 'string',
                value: (_a = preview === null || preview === void 0 ? void 0 : preview.title) !== null && _a !== void 0 ? _a : ''
            },
            url: {
                type: 'string',
                value: (_b = preview.image) !== null && _b !== void 0 ? _b : ''
            }
        };
    }
    if ((imageAssetSource === null || imageAssetSource === void 0 ? void 0 : imageAssetSource.type) === 'cloudinary') {
        return {
            title: {
                type: 'string',
                value: sourceData === null || sourceData === void 0 ? void 0 : sourceData.public_id
            },
            url: {
                type: 'string',
                value: (_e = (_d = (_c = sourceData === null || sourceData === void 0 ? void 0 : sourceData.derived) === null || _c === void 0 ? void 0 : _c[0]) === null || _d === void 0 ? void 0 : _d.secure_url) !== null && _e !== void 0 ? _e : sourceData === null || sourceData === void 0 ? void 0 : sourceData.secure_url
            }
        };
    }
    if ((imageAssetSource === null || imageAssetSource === void 0 ? void 0 : imageAssetSource.type) === 'bynder') {
        // contentful and sanity has own structures and returns fields in the own csi modules
        // git has full response from bynder client so uses this fn
        return {
            title: {
                type: 'string',
                value: sourceData === null || sourceData === void 0 ? void 0 : sourceData.name
            },
            url: {
                type: 'string',
                value: (_g = (_f = sourceData === null || sourceData === void 0 ? void 0 : sourceData.files) === null || _f === void 0 ? void 0 : _f.webImage) === null || _g === void 0 ? void 0 : _g.url
            }
        };
    }
    if ((imageAssetSource === null || imageAssetSource === void 0 ? void 0 : imageAssetSource.type) === 'aprimo') {
        return {
            title: {
                type: 'string',
                value: sourceData === null || sourceData === void 0 ? void 0 : sourceData.title
            },
            url: {
                type: 'string',
                value: (_h = sourceData === null || sourceData === void 0 ? void 0 : sourceData.rendition) === null || _h === void 0 ? void 0 : _h.publicuri
            }
        };
    }
    return {
        title: {
            type: 'string',
            value: ''
        },
        url: {
            type: 'string',
            value: typeof sourceData === 'string' ? sourceData : ''
        }
    };
}
exports.getImageFieldsFromSourceData = getImageFieldsFromSourceData;
function isBuiltInAssetSource(assetSource) {
    return BUILT_IN_ASSET_SOURCES.includes(assetSource);
}
function getAssetSourceBySourceName(assetSources, assetSourceName) {
    if (!assetSourceName) {
        return undefined;
    }
    const assetSource = assetSources.find((assetSources) => assetSources.name === assetSourceName);
    if (!assetSource) {
        // for build-in asset sources, use the name of the source (field[type=image].source) as its type
        if (isBuiltInAssetSource(assetSourceName)) {
            return {
                type: assetSourceName,
                name: assetSourceName
            };
        }
    }
    return assetSource;
}
exports.getAssetSourceBySourceName = getAssetSourceBySourceName;
