// @stackbit/cms-core/dist/utils/meta-fields-filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.META_FIELDS = void 0;
const search_utils_1 = require("./search-utils");
exports.META_FIELDS = {
    createdAt: ({ filter, document }) => ({ field: { type: 'date', value: document.createdAt } }),
    updatedAt: ({ filter, document }) => ({ field: { type: 'date', value: document.updatedAt } }),
    status: ({ filter, document }) => ({ field: { type: 'enum', value: document.status } }),
    hasScheduledActions: ({ filter, document, context }) => {
        var _a;
        return ({
            field: { type: 'boolean', value: !!((_a = context.activeScheduledActionsByDocumentId[document.srcObjectId]) === null || _a === void 0 ? void 0 : _a.length) }
        });
    },
    scheduledActionDate: ({ filter, document, context }) => {
        var _a, _b;
        return {
            result: (_b = (_a = context.activeScheduledActionsByDocumentId[document.srcObjectId]) === null || _a === void 0 ? void 0 : _a.some((scheduledAction) => {
                const field = { type: 'date', value: scheduledAction.executeAt };
                return (0, search_utils_1.isDateFieldMatches)({ field, filter, locale: context.locale });
            })) !== null && _b !== void 0 ? _b : false
        };
    }
};
