// @stackbit/cms-core/dist/utils/store-to-api-docs-converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapStoreAssetsToAPIAssets = exports.mapAssetsToLocalizedApiImages = exports.mapDocumentsToLocalizedApiObjects = void 0;
const lodash_1 = __importDefault(require("lodash"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("@stackbit/utils");
function mapDocumentsToLocalizedApiObjects({ documents, locale, delegate }) {
    return documents.map((document) => documentToLocalizedApiObject({
        document,
        locale,
        delegate
    }));
}
exports.mapDocumentsToLocalizedApiObjects = mapDocumentsToLocalizedApiObjects;
function documentToLocalizedApiObject({ document, locale, delegate }) {
    const { type, fields, getPreview, getDocumentActions, ...rest } = document;
    return (0, utils_1.omitByUndefined)({
        // for historical reasons, stackbit app works with { type: 'object' } not { type: 'document' }
        type: 'object',
        ...rest,
        srcObjectLabel: getPreview({ delegate, locale }).previewTitle,
        actions: getDocumentActions === null || getDocumentActions === void 0 ? void 0 : getDocumentActions(),
        fields: toLocalizedAPIFields({ docFields: fields, delegate, locale })
    });
}
function toLocalizedAPIFields({ docFields, delegate, locale }) {
    return lodash_1.default.mapValues(docFields, (docField) => {
        return (0, utils_1.omitByUndefined)(toLocalizedAPIField({
            docField,
            delegate,
            locale
        }));
    });
}
/**
 * For historical reasons, Stackit app works with fields that can have only one
 * locale. When Stackbit app fetches documents, it passes the current locale.
 * This function converts the internal document field (DocumentField), possibly
 * with nested locales, into an API field (DocumentFieldAPI) with single locale.
 *
 * DocumentField:
 * {
 *   type: "string",
 *   localized: true,
 *   locales: {
 *       'en-US': {
 *           locale: 'en-US',
 *           value: 'hello
 *       },
 *       'fr-FR': {
 *           locale: 'fr-FR',
 *           valie: 'bonjour'
 *       }
 *   }
 * }
 *
 * DocumentFieldAPI:
 * {
 *   type: "string",
 *   localized: true,
 *   locale: 'fr-FR',
 *   value: 'bonjour'
 * }
 *
 * It also does some conversions between field types:
 * type: 'model' => 'object',
 * type: 'reference' => 'unresolved_reference',
 * refType: 'asset' => 'image'
 * refType: 'document' => 'object'
 */
function toLocalizedAPIField({ docField, delegate, locale, isListItem = false }) {
    var _a, _b, _c;
    function localeFields(localized) {
        const isLocalized = !!localized;
        if (isListItem) {
            return null;
        }
        if (!isLocalized) {
            return {
                localized: false
            };
        }
        if (!locale) {
            return null;
        }
        return {
            localized: true,
            locale: locale
        };
    }
    const { getFieldActions, ...restDocField } = docField;
    const fieldActions = getFieldActions === null || getFieldActions === void 0 ? void 0 : getFieldActions({ locale: docField.localized ? locale : undefined });
    docField = restDocField;
    switch (docField.type) {
        case 'string':
        case 'text':
        case 'url':
        case 'slug':
        case 'html':
        case 'number':
        case 'boolean':
        case 'enum':
        case 'date':
        case 'datetime':
        case 'color':
        case 'style': {
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                return {
                    ...base,
                    ...localeProps,
                    ...localeFields(localized),
                    actions: fieldActions
                };
            }
            return {
                ...docField,
                ...localeFields(docField.localized),
                actions: fieldActions
            };
        }
        case 'file':
        case 'json':
        case 'markdown':
        case 'richText': {
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                if (localeProps) {
                    return {
                        ...base,
                        ...localeProps,
                        ...localeFields(localized),
                        actions: fieldActions
                    };
                }
                return {
                    ...base,
                    isUnset: true,
                    ...localeFields(localized),
                    actions: fieldActions
                };
            }
            return {
                ...docField,
                ...localeFields(docField.localized),
                actions: fieldActions
            };
        }
        case 'image': {
            let result;
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                if (localeProps) {
                    const fields = toLocalizedAPIFields({
                        docFields: localeProps.fields,
                        delegate,
                        locale
                    });
                    result = {
                        ...base,
                        ...localeProps,
                        fields,
                        ...localeFields(localized),
                        actions: fieldActions
                    };
                }
                else {
                    result = {
                        ...base,
                        isUnset: true,
                        ...localeFields(localized),
                        actions: fieldActions
                    };
                }
            }
            else {
                if (docField.isUnset) {
                    result = {
                        ...docField,
                        type: 'image',
                        ...localeFields(docField.localized),
                        actions: fieldActions
                    };
                }
                else {
                    const fields = toLocalizedAPIFields({
                        docFields: docField.fields,
                        delegate,
                        locale
                    });
                    result = {
                        ...docField,
                        type: 'image',
                        fields,
                        ...localeFields(docField.localized),
                        actions: fieldActions
                    };
                }
            }
            return result;
        }
        case 'object': {
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                if (localeProps) {
                    const { fields, getPreview, getObjectActions, locale } = localeProps;
                    const objectActions = getObjectActions === null || getObjectActions === void 0 ? void 0 : getObjectActions();
                    return {
                        ...base,
                        ...localeFields(localized),
                        srcObjectLabel: (_a = getPreview({
                            delegate,
                            locale
                        }).previewTitle) !== null && _a !== void 0 ? _a : (isListItem ? 'Item' : 'Object'),
                        actions: concatObjectAndFieldActions(objectActions, fieldActions),
                        fields: toLocalizedAPIFields({
                            docFields: fields,
                            delegate,
                            locale
                        })
                    };
                }
                return {
                    ...base,
                    ...localeFields(localized),
                    actions: fieldActions,
                    isUnset: true
                };
            }
            if (docField.isUnset) {
                return {
                    ...docField,
                    type: 'object',
                    ...localeFields(docField.localized),
                    actions: fieldActions
                };
            }
            const { fields, getPreview, getObjectActions, ...rest } = docField;
            const objectActions = getObjectActions === null || getObjectActions === void 0 ? void 0 : getObjectActions();
            return {
                ...rest,
                type: 'object',
                srcObjectLabel: (_b = getPreview({
                    delegate,
                    locale
                }).previewTitle) !== null && _b !== void 0 ? _b : (isListItem ? 'Item' : 'Object'),
                actions: concatObjectAndFieldActions(objectActions, fieldActions),
                fields: toLocalizedAPIFields({
                    docFields: fields,
                    delegate,
                    locale
                }),
                ...localeFields(docField.localized)
            };
        }
        case 'model': {
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                if (localeProps) {
                    const { fields, getPreview, getObjectActions, ...rest } = localeProps;
                    const objectActions = getObjectActions === null || getObjectActions === void 0 ? void 0 : getObjectActions();
                    return {
                        ...base,
                        ...rest,
                        type: 'object',
                        ...localeFields(localized),
                        srcObjectLabel: getPreview({
                            delegate,
                            locale
                        }).previewTitle,
                        actions: concatObjectAndFieldActions(objectActions, fieldActions),
                        fields: toLocalizedAPIFields({
                            docFields: fields,
                            delegate,
                            locale
                        })
                    };
                }
                return {
                    ...base,
                    type: 'object',
                    isUnset: true,
                    ...localeFields(localized),
                    actions: fieldActions
                };
            }
            if (docField.isUnset) {
                return {
                    ...docField,
                    type: 'object',
                    ...localeFields(docField.localized),
                    actions: fieldActions
                };
            }
            else {
                const { fields, getPreview, getObjectActions, ...rest } = docField;
                const objectActions = getObjectActions === null || getObjectActions === void 0 ? void 0 : getObjectActions();
                return {
                    ...rest,
                    type: 'object',
                    ...localeFields(docField.localized),
                    srcObjectLabel: getPreview({
                        delegate,
                        locale
                    }).previewTitle,
                    actions: concatObjectAndFieldActions(objectActions, fieldActions),
                    fields: toLocalizedAPIFields({
                        docFields: fields,
                        delegate,
                        locale
                    })
                };
            }
        }
        case 'reference': {
            if (docField.localized) {
                const { type, refType, localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                // if reference field isUnset === true, it behaves like a regular object
                if (!localeProps) {
                    return {
                        type: refType === 'asset' ? 'image' : 'object',
                        isUnset: true,
                        ...base,
                        ...localeFields(localized),
                        actions: fieldActions
                    };
                }
                return {
                    type: 'unresolved_reference',
                    refType: refType === 'asset' ? 'image' : 'object',
                    ...base,
                    ...localeProps,
                    ...localeFields(localized),
                    actions: fieldActions
                };
            }
            const { type, refType, ...base } = docField;
            if (base.isUnset) {
                return {
                    type: refType === 'asset' ? 'image' : 'object',
                    ...base,
                    ...localeFields(docField.localized),
                    actions: fieldActions
                };
            }
            return {
                type: 'unresolved_reference',
                refType: refType === 'asset' ? 'image' : 'object',
                ...base,
                ...localeFields(docField.localized),
                actions: fieldActions
            };
        }
        case 'cross-reference': {
            if (docField.localized) {
                const { type, refType, localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                // if reference field isUnset === true, it behaves like a regular object
                if (!localeProps) {
                    return {
                        type: 'object',
                        isUnset: true,
                        ...base,
                        ...localeFields(localized),
                        actions: fieldActions
                    };
                }
                return {
                    type: 'cross-reference',
                    refType: refType,
                    ...base,
                    ...localeProps,
                    ...localeFields(localized),
                    actions: fieldActions
                };
            }
            const { type, refType, ...base } = docField;
            if (base.isUnset) {
                return {
                    type: 'object',
                    ...base,
                    ...localeFields(docField.localized),
                    actions: fieldActions
                };
            }
            return {
                type: 'cross-reference',
                refType: refType,
                ...base,
                ...localeFields(docField.localized),
                actions: fieldActions
            };
        }
        case 'list': {
            if (docField.localized) {
                const { localized, locales, ...base } = docField;
                const localeProps = locales && locale ? locales[locale] : undefined;
                return {
                    ...base,
                    ...localeProps,
                    actions: fieldActions,
                    items: ((_c = localeProps === null || localeProps === void 0 ? void 0 : localeProps.items) !== null && _c !== void 0 ? _c : []).map((field) => toLocalizedAPIField({
                        docField: field,
                        delegate,
                        locale,
                        isListItem: true
                    })),
                    ...localeFields(localized)
                };
            }
            const { items, ...rest } = docField;
            return {
                ...rest,
                ...localeFields(docField.localized),
                actions: fieldActions,
                items: items.map((field) => toLocalizedAPIField({
                    docField: field,
                    delegate,
                    locale,
                    isListItem: true
                }))
            };
        }
        default: {
            const _exhaustiveCheck = docField;
            console.error(`toLocalizedAPIField _exhaustiveCheck failed, docField.type: ${docField['type']}`);
            return _exhaustiveCheck;
        }
    }
}
function mapAssetsToLocalizedApiImages(assets, staticAssetsPublicPath, locale) {
    return assets.map((asset) => assetToLocalizedApiImage(asset, staticAssetsPublicPath, locale));
}
exports.mapAssetsToLocalizedApiImages = mapAssetsToLocalizedApiImages;
function assetToLocalizedApiImage(asset, staticAssetsPublicPath, locale) {
    const { type, fields, ...rest } = asset;
    return {
        type: 'image',
        ...rest,
        fields: localizeAssetFields(fields, staticAssetsPublicPath, locale)
    };
}
function localizeAssetFields(assetFields, staticAssetsPublicPath, locale) {
    var _a, _b, _c, _d, _e, _f;
    const fields = {
        title: {
            type: 'string',
            value: null
        },
        url: {
            type: 'string',
            value: null
        }
    };
    const titleField = assetFields.title;
    if (titleField.localized) {
        if (locale) {
            fields.title = {
                type: 'string',
                value: (_c = (_b = (_a = titleField.locales) === null || _a === void 0 ? void 0 : _a[locale]) === null || _b === void 0 ? void 0 : _b.value) !== null && _c !== void 0 ? _c : null,
                localized: true,
                locale: locale
            };
        }
    }
    else {
        fields.title.value = titleField.value;
    }
    const assetFileField = assetFields.file;
    if (assetFileField.localized) {
        if (locale) {
            const url = (_f = (_e = (_d = assetFileField.locales) === null || _d === void 0 ? void 0 : _d[locale]) === null || _e === void 0 ? void 0 : _e.url) !== null && _f !== void 0 ? _f : null;
            fields.url = {
                type: 'string',
                value: url ? replaceAssetUrlIfNeeded(staticAssetsPublicPath, url) : url,
                localized: true,
                locale: locale
            };
        }
    }
    else {
        fields.url.value = replaceAssetUrlIfNeeded(staticAssetsPublicPath, assetFileField.url);
    }
    return fields;
}
function mapStoreAssetsToAPIAssets(assets, staticAssetsPublicPath, locale) {
    return assets.map((asset) => storeAssetToAPIAsset(asset, staticAssetsPublicPath, locale)).filter((asset) => !!asset);
}
exports.mapStoreAssetsToAPIAssets = mapStoreAssetsToAPIAssets;
function storeAssetToAPIAsset(asset, staticAssetsPublicPath, locale) {
    var _a, _b, _c, _d, _e, _f;
    const assetTitleField = asset.fields.title;
    const localizedTitleField = assetTitleField.localized ? (_a = assetTitleField.locales) === null || _a === void 0 ? void 0 : _a[locale] : assetTitleField;
    const assetFileField = asset.fields.file;
    const localizedFileField = assetFileField.localized ? (_b = assetFileField.locales) === null || _b === void 0 ? void 0 : _b[locale] : assetFileField;
    if (!localizedFileField) {
        return null;
    }
    return {
        objectId: asset.srcObjectId,
        createdAt: asset.createdAt,
        url: (_c = replaceAssetUrlIfNeeded(staticAssetsPublicPath, localizedFileField.url)) !== null && _c !== void 0 ? _c : staticAssetsPublicPath,
        ...(0, utils_1.omitByNil)({
            title: (_d = localizedTitleField === null || localizedTitleField === void 0 ? void 0 : localizedTitleField.value) !== null && _d !== void 0 ? _d : undefined,
            fileName: localizedFileField.fileName,
            contentType: localizedFileField.contentType,
            size: localizedFileField.size,
            width: (_e = localizedFileField.dimensions) === null || _e === void 0 ? void 0 : _e.width,
            height: (_f = localizedFileField.dimensions) === null || _f === void 0 ? void 0 : _f.height
        })
    };
}
function replaceAssetUrlIfNeeded(staticAssetsPublicPath, value) {
    let url = value;
    const normalizedUrl = url === null || url === void 0 ? void 0 : url.toLowerCase();
    if (normalizedUrl && !normalizedUrl.startsWith('http:') && !normalizedUrl.startsWith('https:') && !normalizedUrl.startsWith('//')) {
        url = path_1.default.join(staticAssetsPublicPath, url);
    }
    return url;
}
function concatObjectAndFieldActions(objectActions, fieldActions) {
    let result = [];
    if (fieldActions) {
        result = result.concat(fieldActions);
    }
    if (objectActions) {
        result = result.concat(objectActions);
    }
    if (result.length) {
        return result;
    }
    return undefined;
}
