// @stackbit/cms-contentful/dist/content-poller.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentPoller = void 0;
const lodash_1 = __importDefault(require("lodash"));
const contentful_1 = require("contentful");
const contentful_api_client_1 = require("./contentful-api-client");
class ContentPoller {
    constructor({ spaceId, environment, accessToken, managementToken, pollingIntervalMs = 1000, notificationCallback, logger }) {
        this.logger = logger;
        this.nextSyncToken = null;
        this.running = false;
        this.pollTimeout = null;
        this.pollingIntervalMs = pollingIntervalMs;
        this.notificationCallback = notificationCallback;
        this.client = (0, contentful_1.createClient)({
            space: spaceId,
            environment: environment,
            accessToken: accessToken,
            host: 'preview.contentful.com'
        });
        this.managementClient = (0, contentful_api_client_1.createPlainApiClient)({ accessToken: managementToken, spaceId, environment });
        this.handleTimeout = this.handleTimeout.bind(this);
    }
    start() {
        if (!this.running) {
            this.running = true;
            this.setPollTimeout();
        }
    }
    stop() {
        if (this.pollTimeout) {
            clearTimeout(this.pollTimeout);
            this.pollTimeout = null;
        }
        this.running = false;
    }
    setPollTimeout() {
        this.pollTimeout = setTimeout(this.handleTimeout, this.pollingIntervalMs);
    }
    async handleTimeout() {
        try {
            await this.poll();
        }
        catch (error) {
            this.logger.warn('error polling', { error: error.message });
        }
        if (this.running) {
            this.setPollTimeout();
        }
    }
    async poll() {
        const initial = this.nextSyncToken === null;
        let hasMoreItems = true;
        let hasItems = false;
        const result = {
            entries: [],
            assets: [],
            deletedEntries: [],
            deletedAssets: []
        };
        if (initial) {
            this.logger.info('Running initial sync...');
        }
        while (hasMoreItems) {
            const response = await this.client.sync({
                initial: this.nextSyncToken === null,
                nextSyncToken: this.nextSyncToken,
                resolveLinks: false
            });
            const isEmptyResponse = lodash_1.default.every(lodash_1.default.pick(response, ['entries', 'assets', 'deletedEntries', 'deletedAssets']), lodash_1.default.isEmpty);
            if (this.nextSyncToken === response.nextSyncToken || isEmptyResponse) {
                hasMoreItems = false;
            }
            else {
                if (!initial) {
                    const { entries, assets } = await this.batchRefetchData({
                        entryIds: response.entries.map((entry) => entry.sys.id),
                        assetIds: response.assets.map((entry) => entry.sys.id)
                    });
                    hasItems = true;
                    result.entries = result.entries.concat(entries);
                    result.assets = result.assets.concat(assets);
                    result.deletedEntries = result.deletedEntries.concat(response.deletedEntries);
                    result.deletedAssets = result.deletedAssets.concat(response.deletedAssets);
                }
            }
            this.nextSyncToken = response.nextSyncToken;
        }
        if (!initial && hasItems) {
            this.notificationCallback(result);
        }
        if (initial) {
            this.logger.info('Initial sync done.');
        }
    }
    async batchRefetchData({ entryIds, assetIds }) {
        const limit = 300;
        const entryChunks = lodash_1.default.chunk(entryIds, limit);
        const entries = lodash_1.default.flatMap(await Promise.all(entryChunks.map((chunk) => this.managementClient.entry.getMany({
            query: {
                limit: limit,
                'sys.id[in]': chunk.join(',')
            }
        }))), (result) => result.items);
        const assetChunks = lodash_1.default.chunk(assetIds, limit);
        const assets = lodash_1.default.flatMap(await Promise.all(assetChunks.map((chunk) => this.managementClient.asset.getMany({
            query: {
                limit: limit,
                'sys.id[in]': chunk.join(',')
            }
        }))), (result) => result.items);
        return { entries, assets };
    }
}
exports.ContentPoller = ContentPoller;
