// @stackbit/cms-contentful/src/content-poller.ts
import _ from 'lodash';
import { createClient, ContentfulClientApi, Entry, Asset } from 'contentful';
import { AssetProps, EntryProps, KeyValueMap, PlainClientAPI } from 'contentful-management';
import { Logger } from '@stackbit/types';
import { createPlainApiClient } from './contentful-api-client';

export interface SyncResult {
    entries: EntryProps<any>[];
    assets: AssetProps[];
    deletedEntries: Entry<any>[];
    deletedAssets: Asset[];
}

export type SyncCallback = (result: SyncResult) => void;

export class ContentPoller {
    private readonly logger: Logger;
    private readonly client: ContentfulClientApi;
    private readonly managementClient: PlainClientAPI;
    private readonly notificationCallback: SyncCallback;
    private readonly pollingIntervalMs: number;
    private nextSyncToken: null | string;
    private running: boolean;
    private pollTimeout: NodeJS.Timeout | null;

    constructor({
        spaceId,
        environment,
        accessToken,
        managementToken,
        pollingIntervalMs = 1000,
        notificationCallback,
        logger
    }: {
        spaceId: string;
        environment: string;
        accessToken: string;
        managementToken: string;
        pollingIntervalMs?: number;
        notificationCallback: SyncCallback;
        logger: Logger;
    }) {
        this.logger = logger;
        this.nextSyncToken = null;
        this.running = false;
        this.pollTimeout = null;
        this.pollingIntervalMs = pollingIntervalMs;
        this.notificationCallback = notificationCallback;
        this.client = createClient({
            space: spaceId,
            environment: environment,
            accessToken: accessToken,
            host: 'preview.contentful.com'
        });
        this.managementClient = createPlainApiClient({ accessToken: managementToken, spaceId, environment });
        this.handleTimeout = this.handleTimeout.bind(this);
    }

    start() {
        if (!this.running) {
            this.running = true;
            this.setPollTimeout();
        }
    }

    stop() {
        if (this.pollTimeout) {
            clearTimeout(this.pollTimeout);
            this.pollTimeout = null;
        }
        this.running = false;
    }

    private setPollTimeout() {
        this.pollTimeout = setTimeout(this.handleTimeout, this.pollingIntervalMs);
    }

    async handleTimeout() {
        try {
            await this.poll();
        } catch (error: any) {
            this.logger.warn('error polling', { error: error.message });
        }
        if (this.running) {
            this.setPollTimeout();
        }
    }

    async poll(): Promise<any> {
        const initial = this.nextSyncToken === null;
        let hasMoreItems = true;
        let hasItems = false;
        const result: SyncResult = {
            entries: [],
            assets: [],
            deletedEntries: [],
            deletedAssets: []
        };
        if (initial) {
            this.logger.info('Running initial sync...');
        }
        while (hasMoreItems) {
            const response = await this.client.sync({
                initial: this.nextSyncToken === null,
                nextSyncToken: this.nextSyncToken,
                resolveLinks: false
            });
            const isEmptyResponse = _.every(_.pick(response, ['entries', 'assets', 'deletedEntries', 'deletedAssets']), _.isEmpty);
            if (this.nextSyncToken === response.nextSyncToken || isEmptyResponse) {
                hasMoreItems = false;
            } else {
                if (!initial) {
                    const { entries, assets } = await this.batchRefetchData({
                        entryIds: response.entries.map((entry) => entry.sys.id),
                        assetIds: response.assets.map((entry) => entry.sys.id)
                    });
                    hasItems = true;
                    result.entries = result.entries.concat(entries);
                    result.assets = result.assets.concat(assets);
                    result.deletedEntries = result.deletedEntries.concat(response.deletedEntries);
                    result.deletedAssets = result.deletedAssets.concat(response.deletedAssets);
                }
            }
            this.nextSyncToken = response.nextSyncToken;
        }
        if (!initial && hasItems) {
            this.notificationCallback(result);
        }
        if (initial) {
            this.logger.info('Initial sync done.');
        }
    }

    private async batchRefetchData({
        entryIds,
        assetIds
    }: {
        entryIds: string[];
        assetIds: string[];
    }): Promise<{ entries: EntryProps<KeyValueMap>[]; assets: AssetProps[] }> {
        const limit = 300;
        const entryChunks = _.chunk(entryIds, limit);
        const entries = _.flatMap(
            await Promise.all(
                entryChunks.map((chunk) =>
                    this.managementClient.entry.getMany({
                        query: {
                            limit: limit,
                            'sys.id[in]': chunk.join(',')
                        }
                    })
                )
            ),
            (result) => result.items
        );

        const assetChunks = _.chunk(assetIds, limit);
        const assets = _.flatMap(
            await Promise.all(
                assetChunks.map((chunk) =>
                    this.managementClient.asset.getMany({
                        query: {
                            limit: limit,
                            'sys.id[in]': chunk.join(',')
                        }
                    })
                )
            ),
            (result) => result.items
        );

        return { entries, assets };
    }
}
