// @stackbit/cms-core/dist/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOADER_EXCLUDED_DIRS = exports.LOADER_EXCLUDED_FILES = exports.SIMPLE_VALUE_FIELDS = exports.RICH_TEXT_HINT_MAX_LENGTH = exports.RICH_TEXT_MARK_TYPES = exports.RICH_TEXT_NODE_TYPES = void 0;
var types_1 = require("@stackbit/types");
Object.defineProperty(exports, "RICH_TEXT_NODE_TYPES", { enumerable: true, get: function () { return types_1.RICH_TEXT_NODE_TYPES; } });
Object.defineProperty(exports, "RICH_TEXT_MARK_TYPES", { enumerable: true, get: function () { return types_1.RICH_TEXT_MARK_TYPES; } });
Object.defineProperty(exports, "RICH_TEXT_HINT_MAX_LENGTH", { enumerable: true, get: function () { return types_1.RICH_TEXT_HINT_MAX_LENGTH; } });
exports.SIMPLE_VALUE_FIELDS = ['string', 'text', 'markdown', 'number', 'url', 'slug', 'boolean', 'date', 'datetime', 'color'];
exports.LOADER_EXCLUDED_FILES = [
    'LICENSE.md',
    'README.md',
    'README.theme.md',
    'CONTRIBUTING.md',
    'CHANGELOG.md',
    'stackbit.yaml',
    'netlify.toml',
    'theme.toml',
    'package.json',
    'package-lock.json',
    'yarn-lock.json'
];
exports.LOADER_EXCLUDED_DIRS = ['src/pages'];
