// @stackbit/types/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineStackbitConfig = exports.isDocumentFieldOneOfFieldTypes = exports.isModelFieldSpecificPropsOneOfFieldTypes = exports.isModelFieldOneOfFieldTypes = exports.isOneOfFieldTypes = exports.getLocalizedFieldForLocale = exports.isLocalizedField = exports.getVersion = exports.getInterfaceVersion = void 0;
const path_1 = __importDefault(require("path"));
async function getInterfaceVersion() {
    const packageJson = await Promise.resolve().then(() => __importStar(require(path_1.default.join(__dirname, '../package.json'))));
    return packageJson.version;
}
exports.getInterfaceVersion = getInterfaceVersion;
async function getVersion(options = {}) {
    const interfaceVersion = await getInterfaceVersion();
    if (options.contentSourceVersion) {
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: options.contentSourceVersion
        };
    }
    else if (options.packageJsonPath) {
        const packageJson = await Promise.resolve().then(() => __importStar(require(options.packageJsonPath)));
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: packageJson.version
        };
    }
    else {
        return {
            interfaceVersion: interfaceVersion,
            contentSourceVersion: ''
        };
    }
}
exports.getVersion = getVersion;
function isLocalizedField(field) {
    return !!field.localized;
}
exports.isLocalizedField = isLocalizedField;
/**
 * Returns non-localized version of a DocumentField for the provided `locale`.
 * If the `field` is localized, and the provided locale is not present, this
 * function returns `undefined`. If the `field` is not localized, this function
 * returns the `field` as-is ignoring the `locale`.
 *
 * @example Passing localized field
 * getLocalizedFieldForLocale({
 *   type: 'string',
 *   localized: true,
 *   locales: {
 *       en-US: { value: 'Hello' },
 *       es-ES: { value: 'Hola' },
 *       fr-FR: { value: 'Bonjour' }
 *   }
 * }, 'es-ES');
 * returns:
 * {
 *   type: 'string',
 *   value: 'Hola'
 * }
 *
 * @example Passing non-localized field
 * getLocalizedFieldForLocale({
 *   type: 'string',
 *   value: 'Hello'
 * });
 * returns:
 * {
 *   type: 'string',
 *   value: 'Hello'
 * }
 *
 * @param field
 * @param locale
 */
function getLocalizedFieldForLocale(field, locale) {
    if (field && field.localized) {
        if (!locale) {
            return undefined;
        }
        const { localized, locales, ...base } = field;
        const localizedField = locales === null || locales === void 0 ? void 0 : locales[locale];
        if (!localizedField) {
            return undefined;
        }
        return {
            ...base,
            ...localizedField
        };
    }
    return field;
}
exports.getLocalizedFieldForLocale = getLocalizedFieldForLocale;
function isOneOfFieldTypes(fieldType, fieldTypes) {
    return fieldTypes.includes(fieldType);
}
exports.isOneOfFieldTypes = isOneOfFieldTypes;
function isModelFieldOneOfFieldTypes(modelField, fieldTypes) {
    return fieldTypes.includes(modelField.type);
}
exports.isModelFieldOneOfFieldTypes = isModelFieldOneOfFieldTypes;
function isModelFieldSpecificPropsOneOfFieldTypes(modelField, fieldTypes) {
    return fieldTypes.includes(modelField.type);
}
exports.isModelFieldSpecificPropsOneOfFieldTypes = isModelFieldSpecificPropsOneOfFieldTypes;
function isDocumentFieldOneOfFieldTypes(documentField, fieldTypes) {
    return fieldTypes.includes(documentField.type);
}
exports.isDocumentFieldOneOfFieldTypes = isDocumentFieldOneOfFieldTypes;
function defineStackbitConfig(stackbitConfig) {
    return stackbitConfig;
}
exports.defineStackbitConfig = defineStackbitConfig;
